hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java

```diff
@@ -855,12 +1080,128 @@ public void breakLease(final Path path, final TracingContext tracingContext) thr
+      */
+      List<BlobProperty> srcBlobProperties = getListBlobs(source,
+          tracingContext, null, null, true);
+      final BlobProperty blobPropOnSrc;
+      if (srcBlobProperties.size() > 0) {
+        LOG.debug("src {} exists and is a directory", source);
+        isSrcExist = true;
+        isSrcDir = true;
+        /*
+        * Fetch if there is a marker-blob for the source blob.
+        */
+        BlobProperty blobPropOnSrcNullable = getBlobPropertyWithNotFoundHandling(source, tracingContext);
+        if(blobPropOnSrcNullable == null) {
+          /*
+          * There is no marker-blob, the client has to create marker blob before
+          * starting the rename.
+          */
+          //create marker file; add in srcBlobProperties;
+          LOG.debug("Source {} is a directory but there is no marker-blob", source);
+          azureBlobFileSystem.create(source);
+          Hashtable<String, String> props = new Hashtable<>();
+          props.put(HDI_ISFOLDER, "true");
+          setPathProperties(source, props, tracingContext);
+          blobPropOnSrc = new BlobProperty();
+          blobPropOnSrc.setIsDirectory(true);
+          blobPropOnSrc.setPath(source);
+        } else {
+          LOG.debug("Source {} is a directory but there is a marker-blob", source);
+          blobPropOnSrc = blobPropOnSrcNullable;
+        }
+      } else {
+        LOG.debug("source {} doesn't have any blob in its hierarchy. Checking"
+            + "if there is marker blob for it.", source);
+        blobPropOnSrc = getBlobPropertyWithNotFoundHandling(source, tracingContext);
+        if(blobPropOnSrc != null) {
+          isSrcExist = true;
+          if(blobPropOnSrc.getIsDirectory()) {
+            LOG.debug("source {} is a marker blob", source);
+            isSrcDir = true;
+          } else {
+            LOG.debug("source {} exists but is not a marker blob", source);
+            isSrcDir = false;
+          }
+        } else {
+          LOG.debug("source {} doesn't exist", source);
+          isSrcExist = false;
+          isSrcDir = false;
+        }
+      }
+      srcBlobProperties.add(blobPropOnSrc);
+
+      if (!isSrcExist) {
+        LOG.info("source {} doesn't exists", source);
+        throw new AbfsRestOperationException(HttpURLConnection.HTTP_NOT_FOUND,
+            AzureServiceErrorCode.SOURCE_PATH_NOT_FOUND.getErrorCode(), null, null);
+      }
+      if (isSrcDir) {
+        /*
+        * If source is a directory, all the blobs in the directory have to be
+        * individually copied and then deleted at the source.
+        */
+        LOG.debug("source {} is a directory", source);
+        final RenameAtomicityUtils renameAtomicityUtils;
+        if (isAtomicRenameKey(source.toUri().getPath())) {
+          LOG.debug("source dir {} is an atomicRenameKey", source.toUri().getPath());
+          renameAtomicityUtils = new RenameAtomicityUtils(azureBlobFileSystem,
```

Comment:
Again can we prevent the entire azure blob file system instance to be passed down ?